File: CN/selective_repeat/receiver.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <map>
#include <cstdlib>
#pragma comment(lib, "ws2_32.lib")

using namespace std;

int main() {
    WSADATA wsa;
    WSAStartup(MAKEWORD(2,2), &wsa);

    SOCKET sockfd;
    char buffer[1024];
    sockaddr_in receiverAddr{}, senderAddr{};
    int addr_len = sizeof(senderAddr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    receiverAddr.sin_family = AF_INET;
    receiverAddr.sin_port = htons(5001);
    receiverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    bind(sockfd, (sockaddr*)&receiverAddr, sizeof(receiverAddr));

    int base = 0;
    int windowSize = 4;

    map<int, string> recvBuffer;
    map<int, bool> received; // track all received frames

    cout << "RECEIVER STARTED (SELECTIVE REPEAT)\n\n";

    while (true) {
        memset(buffer, 0, sizeof(buffer));

        recvfrom(sockfd, buffer, sizeof(buffer), 0,
                 (sockaddr*)&senderAddr, &addr_len);

        string frameStr(buffer);
        cout << "RECEIVER: Frame arrived -> " << frameStr << endl;

        int seq = stoi(frameStr.substr(6, frameStr.find(":", 6) - 6));

        // simulate frame loss
        if (rand() % 100 < 20) {
            cout << "RECEIVER: Frame lost\n\n";
            continue;
        }

        // Case 1: Frame inside window
        if (seq >= base && seq < base + windowSize) {

            if (!received[seq]) {
                cout << "RECEIVER: Frame " << seq << " accepted and buffered\n";
                recvBuffer[seq] = frameStr;
                received[seq] = true;
            } else {
                cout << "RECEIVER: Duplicate frame " << seq << endl;
            }

            char ack[20];
            sprintf(ack, "ACK:%d", seq);

            // simulate ACK loss
            if (rand() % 100 < 20) {
                cout << "RECEIVER: ACK lost\n\n";
                continue;
            }

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: ACK sent -> " << ack << endl;

            // deliver in-order frames
            while (received[base]) {
                cout << "RECEIVER: Delivering frame " << base << endl;
                recvBuffer.erase(base);
                base++;
            }
        }

        // Case 2: Old frame (already delivered)
        else if (seq < base) {
            cout << "RECEIVER: Old frame received again\n";

            char ack[20];
            sprintf(ack, "ACK:%d", seq);

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: Re-sent ACK -> " << ack << endl;
        }

        // Case 3: Future frame (outside window)
        else {
            cout << "RECEIVER: Frame outside window ignored\n";
        }

        cout << endl;
    }

    closesocket(sockfd);
    WSACleanup();
    return 0;
}

File: PRAC_EXAM/CN/4_DLL/selective_repeat/receiver.py
import socket
import random

HOST = "127.0.0.1"
PORT = 5200

LOSS_RATE = 0.2
ACK_LOSS_RATE = 0.2
WINDOW_SIZE = 4

server = socket.socket()
server.bind((HOST, PORT))
server.listen(1)

print("RECEIVER STARTED\n")

conn, addr = server.accept()
print("Connected:", addr)

base = 0
received = {}

while True:

    data = conn.recv(1024)

    if not data:
        break

    frame = data.decode()
    print("Received:", frame)

    seq = int(frame.split(":")[1])

    # Simulate frame loss
    if random.random() < LOSS_RATE:
        print("Frame lost\n")
        continue

    # Frame within the receiver's window
    if base <= seq < base + WINDOW_SIZE:

        if not received.get(seq, False):
            print("Buffered frame", seq)
            received[seq] = True

        ack = f"ACK:{seq}"

        # Simulate ACK loss
        if random.random() < ACK_LOSS_RATE:
            print("ACK lost\n")
            continue

        conn.send(ack.encode())
        print("Sent:", ack)

        # Deliver all consecutive frames
        while received.get(base, False):
            print("Delivered frame", base)
            base += 1

    # Duplicate frame
    elif seq < base:
        ack = f"ACK:{seq}"
        conn.send(ack.encode())
        print("Re-sent:", ack)

    print()

conn.close()
server.close()